File: src/ch02/2.5.2_2_TryAsAnExpression1.ts
import * as readline from 'readline';
import { StringDecoder } from 'string_decoder';
function readNumber(reader: readline.Interface) {
const decoder = new StringDecoder('utf8');
reader.question('', (input: string) => {
const number = (() => {
try {
return parseInt(input);
} catch (e) {
if (e instanceof Error && e.name === 'NumberFormatException') {
return null;
}
})();
console.log(number);
});
}
function main() {
const reader = readline.createInterface({
input: process.stdin,
output: process.stdout
});
readNumber(reader);
}
main();

File: src/ch02/2.5.2_2_TryAsAnExpression1.kt
import { BufferedReader, StringReader } from 'some-library';
function readNumber(reader: BufferedReader): void {
    let number: number | null = null;
    try {
        number = parseInt(reader.readLine(), 10);
    } catch (e) {
        if (e instanceof NumberFormatException) {
            number = null;
        } else {
            throw e;
        }
    }
    console.log(number);
}
function main(args: string[]): void {
    const reader = new BufferedReader(new StringReader("not a number"));
    readNumber(reader);
}